chainntnfs/height_hint_cache.go

```diff
@@ -0,0 +1,301 @@
+// have been included within a block.
+type HeightHintCache interface {
+	// CommitSpendHint commits a spend hint for the outpoints to the cache.
+	CommitSpendHint(heightHint uint32, ops ...wire.OutPoint) error
+
+	// QuerySpendHint returns the latest spend hint for an outpoint. The
+	// bool returned signifies whether we have a hint cached for this
+	// outpoint.
+	QuerySpendHint(op wire.OutPoint) (uint32, bool)
+
+	// PurgeSpendHint removes the spend hint for the outpoints from the
+	// cache.
+	PurgeSpendHint(ops ...wire.OutPoint) error
+
+	// CommitConfirmHint commits a confirm hint for the transactions to the
+	// cache.
+	CommitConfirmHint(heightHint uint32, txids ...chainhash.Hash) error
+
+	// QueryConfirmHint returns the latest confirm hint for a transaction.
+	// The bool returned signifies whether we have a hint cached for this
+	// transaction.
+	QueryConfirmHint(txid chainhash.Hash) (uint32, bool)
+
+	// PurgeConfirmHint removes the confirm hint for the transactions from
+	// the cache.
+	PurgeConfirmHint(txids ...chainhash.Hash) error
+}
+
+// heightHintCache is an implementation of the HeightHintCache interface backed
+// by a bolt DB instance where the hints will be stored.
+type heightHintCache struct {
+	db *bolt.DB
+}
+
+// A compile-time check to ensure our HeightHintCache implementation satisfies
+// the interface.
+var _ HeightHintCache = (*heightHintCache)(nil)
+
+// NewHeightHintCache returns a new height hint cache backed by a database.
+func NewHeightHintCache(dbDir string) (HeightHintCache, error) {
+	if err := os.MkdirAll(dbDir, 0700); err != nil {
+		return nil, err
+	}
+
+	db, err := bolt.Open(filepath.Join(dbDir, dbName), dbFilePermission, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	cache := &heightHintCache{db}
+	if err := cache.initBuckets(); err != nil {
+		return nil, err
+	}
+
+	return cache, nil
+}
+
+// initBuckets ensures that the primary buckets used by the circuit are
+// initialized so that we can assume their existence after startup.
+func (c *heightHintCache) initBuckets() error {
+	return c.db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(spendHintBucket)
+		if err != nil {
+			return err
+		}
+
+		_, err = tx.CreateBucketIfNotExists(confirmHintBucket)
+		return err
+	})
+}
+
+// CommitSpendHint commits a spend hint for the outpoints to the cache.
+func (c *heightHintCache) CommitSpendHint(heightHint uint32,
+	ops ...wire.OutPoint) error {
+
+	return c.db.Batch(func(tx *bolt.Tx) error {
+		spendHints := tx.Bucket(spendHintBucket)
+		if spendHints == nil {
+			return ErrCorruptedHeightHintCache
+		}
+
+		var hint bytes.Buffer
+		if err := channeldb.WriteElement(&hint, heightHint); err != nil {
+			return err
+		}
+
+		var outpoint bytes.Buffer
+		for _, op := range ops {
+			outpoint.Reset()
+			err := channeldb.WriteElement(&outpoint, op)
+			if err != nil {
+				return err
+			}
+
+			err = spendHints.Put(outpoint.Bytes(), hint.Bytes())
```

Comment:
are you sure `Put` will work as expected when the slice `key` changes during the transaction. Just make sure we don't run into issues here.

Reply:
Fixed.